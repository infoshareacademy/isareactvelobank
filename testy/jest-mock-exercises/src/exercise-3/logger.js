export const logger = {
    error: (message, error) => console.error('== ups :( ===', message, error)
};