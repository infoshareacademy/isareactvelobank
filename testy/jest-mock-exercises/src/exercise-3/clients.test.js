// Zammockuj getClients i napisz testy!

import { filterClients, getLatestRegisteredClient } from './clients';
import { getClients } from './clients-data';
import { logger } from './logger';
import data from '../../data/db.json'

jest.mock('./clients-data',() => ({
    getClients: jest.fn()
}));

jest.mock('./logger', () => ({
    logger: {
        error: jest.fn()
    }
}))

describe('filterClients', () => {
    afterEach(() => {
        jest.resetAllMocks()
    })

  test('should filter only INTERNAL clients for INTERNAL type', async () => {
    getClients.mockReturnValue(data.clients);
    
    const filteredClients = await filterClients({ type: 'INTERNAL'});
    
    expect(filteredClients).toMatchSnapshot();
  });

  test('should return empty clients when empty list fetched', async () => {
    getClients.mockReturnValue(Promise.resolve([]));
    
    const filteredClients = await filterClients({ type: 'INTERNAL'});
    
    expect(filteredClients).toHaveLength(0);
  });

  test('should return empty list when fetching clients failed', async () => {
    getClients.mockReturnValue(Promise.reject(Error('Failed to fetch')));
    
    const filteredClients = await filterClients({ type: 'INTERNAL'});
   
    expect(filteredClients).toHaveLength(0);
  });

  test('should log error when fetching clients failed', async () => {
    const error = Error('Failed to fetch');
    getClients.mockReturnValue(Promise.reject(error));
    
    await filterClients({ type: 'INTERNAL'});
    
    expect(logger.error).toHaveBeenCalledWith('fetching clients failed!', error);
  });

  test('should not log error when fetching clients successful', async () => {
    getClients.mockReturnValue(Promise.resolve([]));
    
    await filterClients({ type: 'INTERNAL'});
    
    expect(logger.error).not.toHaveBeenCalled();
  });
});
