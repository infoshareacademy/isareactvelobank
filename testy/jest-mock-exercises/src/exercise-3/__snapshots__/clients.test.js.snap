// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterClients should filter only INTERNAL clients for INTERNAL type 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Jan",
    "registered": "2019-05-11T22:00:00.000Z",
    "type": "INTERNAL",
  },
  Object {
    "id": 2,
    "name": "Maria",
    "registered": "2018-04-10T22:00:00.000Z",
    "type": "INTERNAL",
  },
  Object {
    "id": 5,
    "name": "Janina",
    "registered": "2019-12-02T23:00:00.000Z",
    "type": "INTERNAL",
  },
]
`;
